boot/header.S

```diff
@@ -21,63 +41,750 @@
 #define __ASSEMBLY__
 
 #include "boot.h"
+#include "build_version.h"
 #include "peimage.h"
 
 # The EFI loader loads the header at ImageBase, so we have to locate the main program
 # after that. This means we can't load the main program at HIGH_LOAD_ADDR. Pick a load
 # address well away from HIGH_LOAD_ADDR, to avoid overlap when relocating the code.
 
-#define	IMAGE_BASE	0x200000
+#define	IMAGE_BASE		0x200000
+#define	BOOT_PARAMS_START	(SETUP_SECS * 512)
+#define	BOOT_PARAMS_END		(BOOT_PARAMS_START + 4096)
 
-	.section ".header", "ax", @progbits
+	.section ".bootsect", "ax", @progbits
 	.code16
 
 	.globl	boot
 boot:
 	# "MZ", the MS-DOS header signature.
-	.byte 0x4d
-	.byte 0x5a
+	# Fortunately it translates to mostly harmless x86 instructions.
+	.byte 0x4d	# dec	%bp
+	.byte 0x5a	# pop	%dx
 
-        # In case we are booted by a legacy BIOS, print an error message.
-        # Fortunately the MS-DOS header translates to harmless instructions.
+	ljmp	$BOOT_SEG, $init
 
-	ljmp	$BOOT_SEG, $(error - boot)
-error:
+init:
 	movw	%cs, %ax
 	movw	%ax, %ds
 	movw	%ax, %es
 	movw	%ax, %ss
-	xorw	%sp, %sp
+	movw	$BOOT_STACK_TOP, %bp
+	movw	%bp, %sp
 	sti
 	cld
 
-	movw	$error_msg, %si
-0:	lodsb
-	andb	%al, %al
-	jz	wait
-	movb	$0xe, %ah
-	movw	$7, %bx
+	leaw	boot_msg, %bp
+	movw	$(boot_msg_end - boot_msg), %cx
+	call	print_string
+
+	jmp	init_fdd
+
+print_string:
+	# Print string from %es:%bp of length %cx via BIOS routine int 10h: 1301h
+	push	%cx
+	movb	$0x03, %ah		# read cursor pos
+	xorb	%bh, %bh		# page 0
 	int	$0x10
-	jmp	0b
 
-wait:
-	# Allow the user to press a key, then reboot.
-	xorw	%ax, %ax
-	int	$0x16
-	int	$0x19
+	pop	%cx
+	movb	$0x07, %bl		# (page 0 set above) attribute 7 (normal)
+	movw	$0x1301, %ax		# write string, move cursor
+	int	$0x10
+	ret
 
-	# int 0x19 should never return. In case it does, invoke the BIOS.
-        # reset code.
-	ljmp	$0xf000,$0xfff0
+	boot_msg:
+	.ascii	"Loading "
+	boot_msg_end:
 
-        # The PE header pointer.
 	.org	0x3c
+	e_lfanew:
```

Comment:
Is this label used anywhere?

Reply:
Not currently - it is a leftover from my objdump debugging. Deleted, unless you prefer for example to add an assert for it?

Reply:
The .org should fail if there's a problem, so I don't think we need an assert for this.